cpp/src/lists/contains.cu

```diff
@@ -0,0 +1,288 @@
+ * limitations under the License.
+ */
+
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/logical.h>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/contains.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/exec_policy.hpp>
+#include <type_traits>
+
+namespace cudf {
+namespace lists {
+
+namespace {
+
+template <bool search_keys_have_nulls, typename SearchKeyValidityIter>
+std::pair<rmm::device_buffer, size_type> construct_null_mask_impl(
+  cudf::detail::lists_column_device_view const& d_lists, SearchKeyValidityIter search_key_iter)
+{
+  return cudf::detail::valid_if(thrust::make_counting_iterator(0),
+                                thrust::make_counting_iterator(d_lists.size()),
+                                [d_lists, search_key_iter] __device__(auto const& row_index) {
+                                  // Bail, if search_key row is null.
+                                  if (search_keys_have_nulls && !search_key_iter[row_index]) {
+                                    return false;
+                                  }
+                                  auto list = cudf::list_device_view(d_lists, row_index);
+                                  if (list.is_null()) { return false; }
+                                  return thrust::none_of(
+                                    thrust::seq,
+                                    thrust::make_counting_iterator(0),
+                                    thrust::make_counting_iterator(list.size()),
+                                    [&list] __device__(auto const& i) { return list.is_null(i); });
+                                });
+}
+
+/**
+ * @brief Construct null-mask for result of `contains()` called with scalar search key.
+ */
+std::pair<rmm::device_buffer, size_type> construct_null_mask(
+  cudf::detail::lists_column_device_view const& d_lists,
+  cudf::scalar const& search_key,
+  bool input_column_has_nulls,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  using namespace cudf;
+  using namespace cudf::detail;
+
+  if (search_key.is_valid(stream) && !input_column_has_nulls) {
+    return std::make_pair(rmm::device_buffer{0, stream, mr}, size_type{0});
+  }
+
+  if (!search_key.is_valid(stream)) {
+    return std::make_pair(cudf::create_null_mask(d_lists.size(), mask_state::ALL_NULL, mr),
+                          d_lists.size());
+  }
+
+  return construct_null_mask_impl<false>(d_lists, thrust::make_constant_iterator(true));
+}
+
+/**
+ * @brief Construct null-mask for result of `contains()` called with column of search keys.
+ */
+std::pair<rmm::device_buffer, size_type> construct_null_mask(
```

Comment:
Are you sure these two functions can't be combined by replacing the search keys parameter with an iterator parameter at a higher level?

Reply:
Sorry, I been trying, and I still don't see a good way. 

For instance, the `scalar` overload has a short-circuit path for when the `search_key` itself is null. Forgoing this would introduce an otherwise avoidable check in `construct_null_mask_impl()`.

Reply:
Is doing the check in the impl function a problem?

Reply:
> Is doing the check in the impl function a problem?

Hmm... Potentially not. Will update.

Reply:
I have combined the two functions as per your suggestion. This has helped collapse the two detail APIs for `collect()` as well.